### DISCUSSION TITLE
Converting to Cloud Optimized GeoTiff (COG) format

### DISCUSSION THREAD

**Discussion by RichardScottOZ** [banesullivan/localtileserver#27] [Ideas]
Maybe a paragraph in the README 'here's some things to make COGs with'?

**Comment by banesullivan** [comment]
Cloud optimized geotiffs are really only necessary when accessing files over in the internet (passing a URL and it going through GDALs VFS interface). If the file is on your local system, it being a COG isn't all that necessary. 

With that said, large image's `large-image-converter` provides a `convert()` to do the conversion very easily. https://pypi.org/project/large-image-converter/

**Comment by RichardScottOZ** [comment]
Yes, I have terabytes of the things. :)  Good suggestion for others.  
This script can be handy:- https://github.com/rouault/cog_validator

```python
import rioxarray
kw = {}
kw.update(driver="COG")
result['red'].rio.to_raster('ResultRed.tif', **kw)
```

**Reply by banesullivan** [reply]
I have not tested the above conversion, but I'm skeptical of it. Just because something is a COG doesn't necessarily mean it is pyramidal or properly tiled. An improperly tiled COG is no more efficient than a non-COG geotiff when reading over the VFS interface.

I'd recommend using GDAL directly to create the COGs as you need to make sure you have a few additional options enabled beyond just the COG driver. For example:

```py
options = [
         '-co',
         'COMPRESS=LZW',
         '-co',
         'PREDICTOR=YES',
         '-of',
         'COG',
         '-co',
         'BLOCKSIZE=256',
     ]

ds = gdal.Open(src_path)
ds = gdal.Translate(output_path, ds, options=options)
```
or from the command line:

```bash
gdal_translate <input> <output>.tiff -of COG -co BIGTIFF=IF_SAFER -co BLOCKSIZE=256 -co COMPRESS=LZW
```

This is exactly what is implemented in `large_image_converter`'s `convert()` function by default (along with some other options that are useful): https://github.com/girder/large_image/blob/4e928166f228fe894c38e4b01af5370e72f7229c/large_image/tilesource/utilities.py#L208

So you can do this by:

```py
import large_image_converter
large_image_converter.convert(str(input_path), str(output_path))
```

from https://pypi.org/project/large-image-converter/

`large_image` and `large_image_source_gdal` are a dependencies of `localtileserver` but `large_image_converter` is not and requires having `libtiff` installed as well (like GDAL, conda is your best friend here when it comes to installing)


------

I will see if I can find some time to do a full write up / blog post on this because its a rather common problem

**Reply by Boorhin** [reply]
I have tried that and I receive this from large image converter
I guess there is something not right in the requirements
```
import large_image_converter

large_image_converter.convert('rio_stacked_geotiff_test.tiff', 'Staked_COG.tiff')

Generating '/opt/conda/lib/python3.7/site-packages/libtiff/tiff_h_4_3_0.py' from '/opt/conda/lib/../include/tiff.h'

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
/tmp/ipykernel_18599/2711869581.py in <module>
----> 1 import large_image_converter
      2 large_image_converter.convert('rio_stacked_geotiff_test.tiff', 'Staked_COG.tiff')

/opt/conda/lib/python3.7/site-packages/large_image_converter/__init__.py in <module>
     18 import large_image
     19 
---> 20 from . import format_aperio
     21 
     22 pyvips = None

/opt/conda/lib/python3.7/site-packages/large_image_converter/format_aperio.py in <module>
      2 import math
      3 
----> 4 import large_image_source_tiff
      5 import tifftools
      6 

/opt/conda/lib/python3.7/site-packages/large_image_source_tiff/__init__.py in <module>
     33 from large_image.tilesource import FileTileSource, nearPowerOfTwo
     34 
---> 35 from .tiff_reader import (InvalidOperationTiffException, IOTiffException,
     36                           IOTiffOpenException, TiffException,
     37                           TiledTiffDirectory, ValidationTiffException)

/opt/conda/lib/python3.7/site-packages/large_image_source_tiff/tiff_reader.py in <module>
     33 
     34 try:
---> 35     from libtiff import libtiff_ctypes
     36 except ValueError as exc:
     37     # If the python libtiff module doesn't contain a pregenerated module for

/opt/conda/lib/python3.7/site-packages/libtiff/__init__.py in <module>
     33 _sys.path.append(_os.path.join(_os.path.dirname(__file__),
     34                                'bitarray-a1646c0'))
---> 35 import bitarray as _bitarray    # noqa: F402
     36 _sys.modules['libtiff.bitarray'] = _bitarray

ModuleNotFoundError: No module named 'bitarray'
```

**Reply by banesullivan** [reply]
Huh, that's odd. We may want to open an issue in pylibtiff as the issue appears to be coming from there

**Reply by banesullivan** [reply]
Ref https://github.com/pearu/pylibtiff/issues/108

**Reply by Boorhin** [reply]
OK I come back to this as I have tried another method to generate a COG. 
I just want to point out first that the PREDICATOR=YES cannot work, it takes the value 1, 2 or 3 and COG is not an option according to gdal doc
https://gdal.org/drivers/raster/gtiff.html#raster-gtiff
Also BLOCKSIZE is by default 256
Maybe it is because of different versions?

**Comment by RichardScottOZ** [comment]
I have found it to be pretty good generally.  Could be issues with dask, maybe, but the vanilla version:-

![image](https://user-images.githubusercontent.com/72196131/144736218-ab87e0bd-1078-4732-b8f4-e801064ab688.png)

**Reply by RichardScottOZ** [reply]
As in hasn't failed in the 'gdal validator says it is ok' sense, has overviews, etc.

e.g. rioxarray - rasterio - gdal - only thing that could break that is dask, [assuming just using geotiffs]  I think, if using chunks?

**Comment by RichardScottOZ** [comment]
Although had reason to use the above in an environment with only GDAL 3.0.

```python
session = boto3.Session(profile_name='cogstuff')
with rasterio.Env(session):

    s3 = session.resource('s3')

    object = s3.Object('cogbucket', 'test/test_COG_2.tif')

    tif_file2 = '/vsis3/cogbucket/test/test.tif'
    ds = gdal.Open(tif_file2  )
    print(ds)

    gdal.Translate('/vsimem/temp_file.tif', ds, format="GTiff", creationOptions=['TILED=YES' 'COPY_SRC_OVERVIEWS=YES' 'COMPRESS=LZW', 'BLOCKXSIZE=512', 'BLOCKYSIZE=512', 'PREDICTOR=YES'])

    f = gdal.VSIFOpenL('/vsimem/temp_file.tif', 'rb')
    gdal.VSIFSeekL(f, 0, 2)  # seek to end
    size = gdal.VSIFTellL(f)
    gdal.VSIFSeekL(f, 0, 0)  # seek to beginning
    data = gdal.VSIFReadL(1, size, f)
    gdal.VSIFCloseL(f)

    object.put(Body=data)

    ds = None
```

**Comment by Boorhin** [comment]
I have an issue with displaying properly plots from xarray (stored in zarr) through mapbox (don't really know how I would tile the underlaying raster) as I am running into memory issues.
So I was wondering if it could be efficient to make COG on the fly? 
My issue is that I need to stack the rasters according to user selections. 
Or I could maybe make a single COG with multiple layers but is that easy to stack? I am also a bit unsure on how to do that. 
At the moment I read the part of the different zarr archives I need (that's several datasets) and I stack all the zarr into a single xarray dataarray thanks to datashader and I create the raster that is overlaid by ScatterMapBox in Plotly. however, the image I create, because of our expanding computations start to be unmanageable online and would need tiling/pyramidal compression.

**Reply by Boorhin** [reply]
I think I got the answer to my question here ...
https://docs.digitalearthafrica.org/en/latest/sandbox/notebooks/Frequently_used_code/Exporting_GeoTIFFs.html

**Comment by banesullivan** [comment]
FWIW, I'm working on a blog (will share when it posts) to highlight differences in "valid" COGs. Here is a script that will highlight the difference for the same image saved as two formats:

1. COG without overviews (invalid)
2. COG with overviews (valid)

```py
from memory_profiler import profile
from osgeo import gdal

kwargs = dict(
    xRes=611.49622628141,
    yRes=611.49622628141,
    outputBounds=[-9079495.967826376, 3443946.7464169012, -8922952.933898335, 3600489.780344942],
    srcSRS="+proj=utm +zone=17 +datum=WGS84 +units=m +no_defs",
    dstSRS="epsg:3857",
)


@profile
def get_tile(path):
    # Open from remote server that supports HTTP range requests
    source = gdal.Open(path)
    # Extract tile 8/70/105
    ds = gdal.Warp("", source, format="VRT", **kwargs)
    # Perform operation and fetch data
    return ds.ReadAsArray()


valid = "/vsicurl?url=https%3A%2F%2Fdata.kitware.com%2Fapi%2Fv1%2Ffile%2F62d70e66bddec9d0c478d85c%2Fdownload&use_head=no&list_dir=no"
invalid = "/vsicurl?url=https%3A%2F%2Fdata.kitware.com%2Fapi%2Fv1%2Ffile%2F626733ce4acac99f42fa5894%2Fdownload&use_head=no&list_dir=no"

# Valid COG
get_tile(valid)
# uses ~25Mb
# 13.8 ms ± 1.54 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

# Non-COG
get_tile(invalid)
# Uses ~160 Mb
# 17.2 s ± 1.74 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

The script and data are hosted here: https://data.kitware.com/#folder/6267337e4acac99f42fa5844

The results are pretty compelling:
![Screen Shot 2022-07-19 at 2 22 02 PM](https://user-images.githubusercontent.com/22067021/179841275-0c0cf7b5-0ef8-450f-858a-bfe69a2bb670.png)

**Reply by Boorhin** [reply]
I just wanted to come back on testing datacube write_cog utility.
I have not managed to export the multilayered xr.dataArray into a COG the kernel crashes
can I directly export a rio.dataarray into COG with gdal?

**Comment by RichardScottOZ** [comment]
Can you add your code?

**Reply by Boorhin** [reply]
Ok, I have tried a lot of different things with more or less success. My objective is to take a xr.Dataset representing different parameters of a regular grid with different variable names into a COG with multiple layers corresponding to these parameters.
What worked was similar to when I was using datashader to make png of the datasets but in more tedious and slower

``` python
import gcsfs
import xarray as xr
import rioxarray

fs= gcsfs.GCSFileSystem()
bucket='sealice_db/westcoast_master_50m.zarr/'
mapped=gcsfs.mapping.GCSMap(bucket, gcs=fs, check=True, create=False)
masta=xr.open_zarr(mapped)
```
masta

xarray.Dataset

    Dimensions:
        y: 8389x: 18953
    Coordinates:
        x
        (x)
        float64
        -9.086e+05 -9.085e+05 ... 7.46e+05
        y
        (y)
        float64
        8.13e+06 8.13e+06 ... 7.318e+06
    Data variables: (125)
    Attributes: (0)

``` python
temp_arr= masta.to_stacked_array('v', ['y', 'x']).sum(dim='v').rio.write_crs('EPSG:3857')
mk_geotiff=  temp_arr.rio.to_raster('rio_stacked_geotiff_test.tiff', driver='COG')
import large_image_converter
large_image_converter.convert('rio_stacked_geotiff_test.tiff', 'Staked_COG.tiff')
``` 

The first geotiff I obtain is massive and very slow to produce whereas the converted one is lighter, faster to produce and according to Bane, a real/good COG.
``` bash
$ ls -l *.tiff
-rw-r--r-- 1 jupyter jupyter  146304333 Oct 27 17:23 Staked_COG.tiff
-rw-r--r-- 1 jupyter jupyter 1868578332 Oct 27 17:05 rio_stacked_geotiff_test.tiff
``` 

So I could do with doing the stacking and parametrisation of each layers with xarray as it is very fast but I would need to use a similar method than the large-image-converter that eats directly a xarray DataArray and spits out a proper COG. I was under the impression that datashader was capable of it but they must have the worst documentation, I cannot follow it.

**Comment by RichardScottOZ** [comment]
If you are reading zarr and hence getting a dask backed dataset - you will have problems writing those via rioxarray using the COG driver.   At least last tiem It ried it. You'd have to definite it with the various rasterio parameters as Bane gives above.

A 10K x 20K raster will take a while if you are downloading it on home type bandwidth.

**Reply by Boorhin** [reply]
Hi Richard,
This is all made directly on the cloud in a VM where the data are hosted so reading data is very very fast. I have done it with Datashader and it worked pretty well but the display is not optimised because it only does png. This is for an app deployment, the point is to render the image interactively.

**Reply by Boorhin** [reply]
For the parameters, when you add the crs, rioxarray should have everything it needs (I assume)

**Reply by Boorhin** [reply]
FYI if you want to test the same thing with Datashader, this will output a png and be very fast
``` python
from datashader import transfer_functions as tf
tf.shade(temp_arr).to_pil()
```

**Comment by Boorhin** [comment]
I have tried a new method to generate a "COG" directly from a stacked xarray dataset. Again it is quite slow and I have difficulties to understand. It seems to me that all the formats are similar so there should not be a lot of operations. The data go in memory and it seems to run for a while with quite a good usage of the cores but I am puzzled that it takes so much. I must be doing something wrong
``` python
from osgeo import osr, gdal
import xarray as xr


ds= xr.open_zarr('westcoast_master_50m.zarr')
temp_arr= ds.to_stacked_array('v', ['y', 'x']).sum(dim='v')

res=50
width, height=len(ds.x), len(ds.y)

COG='Optimised Geotiff.tif'

options = [
         'TILED=YES',
         'COMPRESS=LZW',
         'COPY_SRC_OVERVIEWS=YES',
     ]

driver=gdal.GetDriverByName('GTiff')

outraster=driver.Create(COG, width, height,1, gdal.GDT_Float32, options=options)
outraster.SetGeoTransform([ds.x.values[0],50,0,ds.y.values[0],-50,0])
outrasterSRS= osr.SpatialReference()
outrasterSRS.ImportFromEPSG(3857)
outraster.SetProjection(outrasterSRS.ExportToWkt())
outraster.GetRasterBand(1).WriteArray(temp_arr.values) #<-- heavy stuff in memory
outraster.FlushCache()
```
This time I got a 56.4 Mb Geotiff

**Reply by Boorhin** [reply]
This way of making geotiff didn't work I get some bonkers values and no data

> ><h1 style="color: rgb(88, 150, 50); font-size: x-large; font-weight: normal; background-image: none; background-color: rgb(255, 255, 255); padding: 0.75em 0px 0px; margin: 0px; line-height: inherit; font-family: Lato, 'Open Sans', 'Lucida Grande', 'Segoe UI', Arial, sans-serif; font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-tap-highlight-color: rgba(0, 0, 0, 0.4); -webkit-text-stroke-width: 0px; background-position: initial initial; background-repeat: initial initial;">Information from provider</h1><hr style="border-width: 1px 0px 0px; height: 0px; border-top-style: solid; border-top-color: black; color: rgb(0, 0, 0); font-family: Lato, 'Open Sans', 'Lucida Grande', 'Segoe UI', Arial, sans-serif; font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-tap-highlight-color: rgba(0, 0, 0, 0.4); -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255);">
> 
> Extent | -908621.4786600348306820,7889975.3941366896033287 : -668621.4786600348306820,8129975.3941366896033287
> -- | --
> Width | 4800
> Height | 7905
> Data type | Float32 - Thirty two bit floating point
> GDAL Driver Description | GTiff
> GDAL Driver Metadata | GeoTIFF
> Dataset Description | ...app/data/Optimised Geotiff.tif
> Compression | LZW
> Band 1 | STATISTICS_APPROXIMATE=YESSTATISTICS_MAXIMUM=0STATISTICS_MEAN=0STATISTICS_MINIMUM=0STATISTICS_STDDEV=0STATISTICS_VALID_PERCENT=100
> More information | AREA_OR_POINT=Area
> Dimensions | X: 4800 Y: 7905 Bands: 1
> Origin | -908621.4786600348306820,8129975.3941366896033287
> Pixel Size | 50,0
